jsx/Form.js

```diff
@@ -134,6 +134,138 @@ var FormElement = React.createClass({
+
+  getKeyFromValue: function(value) {
+    var options = this.props.options;
+    var keyValue = Object.keys(options).find(function(o) {
+      return options[o] === value;
+    });
+
+    return keyValue;
+  },
+
+  handleChange: function(e) {
+    var value = this.getKeyFromValue(e.target.value);
+    // if not in strict mode and key value is not defined (i.e., not in options)
+    // set value equal to e.target.value
+    if (!this.props.strictDatalist && value === undefined) {
+      value = e.target.value;
+    }
+    this.props.onUserInput(this.props.name, value);
+  },
+
+  handleBlur: function(e) {
+    // null out entry if not present in options in strict mode
+    if (this.props.strictDatalist) {
+      var value = e.target.value;
+      var options = this.props.options;
+      if (Object.values(options).indexOf(value) === -1) {
+        // empty string out both the hidden value as well as the input text
+        document.querySelector(`input[name="${this.props.name + '_list'}"]`).value = '';
+        this.props.onUserInput(this.props.name, '');
+      }
+    }
+  },
+
+  render: function() {
+    var required = this.props.required ? 'required' : null;
+    var disabled = this.props.disabled ? 'disabled' : null;
+    var options = this.props.options;
+    var strictMessage = 'Entry must be included in provided list of options.';
+    var errorMessage = null;
+    var requiredHTML = null;
+    var elementClass = 'row form-group';
+
+    // Add required asterix
+    if (required) {
+      requiredHTML = <span className="text-danger">*</span>;
+    }
+
+    // Add error message
+    if (this.props.hasError || (this.props.required && this.props.value === "")) {
+      var msg = this.props.errorMessage;
+      msg += (this.props.strictDatalist ? ' ' + strictMessage : '');
+      errorMessage = <span>{msg}</span>;
+      elementClass = 'row form-group has-error';
+    } else if (this.props.strictDatalist && this.props.value === "") {
+      errorMessage = <span>{strictMessage}</span>;
+      elementClass = 'row form-group has-error';
+    }
+
+    return (
+      <div className={elementClass}>
+        <label className="col-sm-3 control-label" htmlFor={this.props.label}>
+          {this.props.label}
+          {requiredHTML}
+        </label>
+        <div className="col-sm-9">
+          <input
+            type="text"
+            name={this.props.name + '_list'}
+            id={this.props.id}
+            list={this.props.name + '_list'}
+            className="form-control"
+            disabled={disabled}
+            onChange={this.handleChange}
+            onBlur={this.handleBlur}
+          />
+          <datalist id={this.props.name + '_list'}>
+            {Object.keys(options).map(function(option) {
+              return (
+                <option value={options[option]} key={option}/>
+              );
+            })}
+          </datalist>
+          <input type="hidden" name={this.props.name} value={this.props.value}/>
```

Comment:
what's this hidden input element for?

Reply:
basically the text input is acting as a facade, and the actual submitted value is getting stored here

Reply:
Correct me if I'm wrong @davidblader, but _stored_ might not be the best word in this case. The first `input` works together with the `datalist` to present an input field to the user that can be searched or selected from the given options (refer to https://www.w3schools.com/tags/tag_datalist.asp). However, when a value is selected, it is submitted through the second `input` of type `hidden`. This is done to allow the `datalist` to function using key-value pairs, wherein the `value` is presented as a `label` to the user in the first `input`, and the `key` is submitted to second `input` (refer to top answer: https://stackoverflow.com/questions/29882361/show-datalist-labels-but-submit-the-actual-value).

Reply:
technically the hidden input isn't even required since the form data which gets sent to the server is usually built based on the formData object in `state`, which is set by onUserInput, but i am so used to thinking of each entry in formData having a corresponding DOM input element that i added it for my own personal comfort

Reply:
Wait, so if it's not necessary should it be removed?

Reply:
¯\_(ツ)_/¯

Reply:
@davidblader that's what I thought but fair enough! i feel like to keep consistent with the  other elements in Form.js, and if the hidden input element value isn't ever used, that we shouldn't include this line. i'm not totally against it though

Reply:
after some thought i think it's best to keep the hidden input, in case the submitted value needs to be retrievable from another external client-side script

Reply:
I think it should be taken out if it's not used. DOM access is slow, and if it needs to be retrievable there should be some other, more explicit/controlled way of doing it.